lib/public_key/src/pubkey_ocsp.erl

```diff
@@ -153,31 +157,91 @@ get_nonce_value([#'Extension'{
-            {error, Reason}
+verify_signature(_, _, _, [], _, _, _) ->
+    {error, ocsp_responder_cert_not_found};
+verify_signature(ResponseDataDer, SignatureAlgo, Signature,
+                 [ResponderCert | RCs], ResponderID, IssuerCert,
+                 IsTrustedResponderFun) ->
+    maybe
+        true ?= is_responder_cert(ResponderID, ResponderCert),
+        true ?= is_authorized_responder(ResponderCert, IssuerCert,
+                                        IsTrustedResponderFun),
+        ok ?= do_verify_signature(ResponseDataDer, Signature, SignatureAlgo,
+                                  ResponderCert)
+    else
+        _->
+            verify_signature(ResponseDataDer, SignatureAlgo, Signature,
+                             RCs, ResponderID, IssuerCert,
+                             IsTrustedResponderFun)
     end.
 
-find_responder_cert(_ResponderID, []) ->
-    {error, ocsp_responder_cert_not_found};
-find_responder_cert(ResponderID, [Cert | TCerts]) ->
-    case is_responder(ResponderID, Cert) of
+verify_past_timestamp(Timestamp) ->
+    {Now, TimestampSec} = get_time_in_sec(Timestamp),
+    verify_timestamp(Now, TimestampSec, past_timestamp).
+
+verify_future_timestamp(Timestamp) ->
+    {Now, TimestampSec} = get_time_in_sec(Timestamp),
+    verify_timestamp(Now, TimestampSec, future_timestamp).
+
+verify_timestamp(Now, Timestamp, past_timestamp) when Timestamp =< Now ->
+    ok;
+verify_timestamp(Now, Timestamp, future_timestamp) when Now =< Timestamp ->
+    ok;
+verify_timestamp(_, _, _) ->
+    {error, ocsp_stale_response}.
+
+get_time_in_sec(Timestamp) ->
+    Now = calendar:datetime_to_gregorian_seconds(calendar:universal_time()),
+    TimestampSec = pubkey_cert:time_str_2_gregorian_sec(
+                     {generalTime, Timestamp}),
+    {Now, TimestampSec}.
+
+verify_next_update(asn1_NOVALUE) ->
+    ok;
+verify_next_update(NextUpdate) ->
+    verify_future_timestamp(NextUpdate).
+
+is_responder_cert({byName, Name}, #cert{otp = Cert}) ->
+    public_key:der_encode('Name', Name) == get_subject_name(Cert);
+is_responder_cert({byKey, Key}, #cert{otp = Cert}) ->
+    Key == crypto:hash(sha, get_public_key(Cert)).
+
+is_authorized_responder(CombinedResponderCert = #cert{otp = ResponderCert},
+                        IssuerCert, IsTrustedResponderFun) ->
+    Case1 =
+        %% the CA who issued the certificate in question signed the
+        %% response
+        fun() ->
+                ResponderCert == IssuerCert
+        end,
+    Case2 =
+        %% a CA Designated Responder (Authorized Responder, defined in
+        %%      Section 4.2.2.2) who holds a specially marked certificate
+        %%      issued directly by the CA, indicating that the responder may
+        %%      issue OCSP responses for that CA (id-kp-OCSPSigning)
+        fun() ->
+                public_key:pkix_is_issuer(ResponderCert, IssuerCert) andalso
+                                 designated_for_ocsp_signing(ResponderCert)
```

Comment:
I think we want to verify the signature here also, that is public_key:pkix_verify(ResponderCert, IssuerCertsPublicKey) not only see that the issuer name check out!

Reply:
Maybe this is actually done further up!

Reply:
Yes. This call realizes this:
`ok ?= do_verify_signature(ResponseDataDer, Signature, SignatureAlgo, ResponderCert)`

OCSP response contains list of certificates, we need to iterate over that list.
Once a potential candidate is found, signature of response is verified by `do_verify_signature` call.